### PR TITLE
HIVE-24738: Reuse committed filelist from directinsert manifest during loadPartition

### PR DESCRIPTION
**Pull request by pvargacl**
### What changes were proposed in this pull request?
Reuse committed file list from directinsert manifest file during loadPartition to avoid FileSystem listing

### Why are the changes needed?
Performance improvement

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
Existing unit tests.

### CODE REVIEW

File: ql/src/java/org/apache/hadoop/hive/ql/exec/Utilities.java [hunk 1]
```diff
@@ -2862,29 +2862,29 @@ public boolean skipProcessing(Task<?> task) {
...
-        for (String partSpec : dynamicPartitionSpecs) {
-          allPartition.add(new Path(loadPath, partSpec));
+        for (Map.Entry<String, List<Path>> partSpec : dynamicPartitionSpecs.entrySet()) {
+          allPartition.put(new Path(loadPath, partSpec.getKey()), partSpec.getValue());
         }
       } else {
         List<FileStatus> status = HiveStatsUtils.getFileStatusRecurse(loadPath, numDPCols, fs);
         for (FileStatus fileStatus : status) {
-          allPartition.add(fileStatus.getPath());
+          allPartition.put(fileStatus.getPath(), null);
```
Comment (deniskuzZ) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/Utilities.java:2879]: Would it require much of refactor if we switch to Optional here?

File: ql/src/java/org/apache/hadoop/hive/ql/metadata/Hive.java [hunk 2]
```diff
@@ -2362,10 +2362,17 @@ private Partition loadPartitionInternal(Path loadPath, Table tbl, Map<String, St
...
         if (newFiles != null) {
-          newFileStatuses = listFilesCreatedByQuery(loadPath, writeId, stmtId);
-          newFiles.addAll(newFileStatuses.stream()
-              .map(FileStatus::getPath)
-              .collect(Collectors.toList()));
+          if (!newFiles.isEmpty()) {
+            // We already know the file list from the direct insert manifestfile
+            FileSystem srcFs = loadPath.getFileSystem(conf);
+            newFileStatuses = new ArrayList<>();
+            for (Path filePath : newFiles) {
```
Comment (deniskuzZ) [inline ql/src/java/org/apache/hadoop/hive/ql/metadata/Hive.java:2369]: if you already have streams here, could we use foreach?

File: ql/src/java/org/apache/hadoop/hive/ql/metadata/Hive.java [hunk 3]
```diff
@@ -2362,10 +2362,17 @@ private Partition loadPartitionInternal(Path loadPath, Table tbl, Map<String, St
...
+              newFileStatuses.add(srcFs.getFileStatus(filePath));
+            }
+          } else {
+            newFileStatuses = listFilesCreatedByQuery(loadPath, writeId, stmtId);
+            newFiles.addAll(newFileStatuses.stream().map(FileStatus::getPath).collect(Collectors.toList()));
```
Comment (deniskuzZ) [inline ql/src/java/org/apache/hadoop/hive/ql/metadata/Hive.java:2374]: can you please format it for better readability

File: ql/src/java/org/apache/hadoop/hive/ql/exec/Utilities.java [hunk 4]
```diff
@@ -4533,6 +4537,12 @@ public static void handleDirectInsertTableFinalPath(Path specPath, String unionS
           if (!committed.add(path)) {
             throw new HiveException(nextFile + " was specified in multiple manifests");
           }
+          for (Map.Entry<String, List<Path>> dynpath : dynamicPartitionSpecs.entrySet()){
```
Comment (deniskuzZ) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/Utilities.java:4540]: why not just filter and collect?

File: ql/src/java/org/apache/hadoop/hive/ql/exec/Utilities.java [hunk 5]
```diff
@@ -4521,7 +4523,9 @@ public static void handleDirectInsertTableFinalPath(Path specPath, String unionS
           for (int i = 0; i < partitionCount; ++i) {
             String nextPart = mdis.readUTF();
             Utilities.FILE_OP_LOGGER.debug("Looking at dynamic partition {}", nextPart);
-            dynamicPartitionSpecs.add(nextPart);
+            if (!dynamicPartitionSpecs.containsKey(nextPart)) {
+              dynamicPartitionSpecs.put(nextPart, new ArrayList<>());
```
Comment (pvary) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/Utilities.java:4527]: nit: why List?

File: ql/src/java/org/apache/hadoop/hive/ql/plan/FileSinkDesc.java [hunk 6]
```diff
@@ -106,7 +108,7 @@
 
   private Set<FileStatus> filesToFetch = null;
 
-  private Set<String> dynPartitionValues = new HashSet<>();
+  private Map<String, List<Path>> dynPartitionValues = new HashMap<>();
```
Comment (pvary) [inline ql/src/java/org/apache/hadoop/hive/ql/plan/FileSinkDesc.java:111]: nit: a comment would be nice describing keys and values

File: ql/src/java/org/apache/hadoop/hive/ql/exec/Utilities.java (hunk 1 above)
Comment (pvargacl) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/Utilities.java]: done

File: ql/src/java/org/apache/hadoop/hive/ql/metadata/Hive.java (hunk 3 above)
Comment (pvargacl) [reply ql/src/java/org/apache/hadoop/hive/ql/metadata/Hive.java]: done

File: ql/src/java/org/apache/hadoop/hive/ql/metadata/Hive.java (hunk 2 above)
Comment (pvargacl) [reply ql/src/java/org/apache/hadoop/hive/ql/metadata/Hive.java:2371]: getFileStatus throws IOException, and I do not want to wrap ip

File: ql/src/java/org/apache/hadoop/hive/ql/exec/Utilities.java (hunk 4 above)
Comment (pvargacl) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/Utilities.java]: done

File: ql/src/java/org/apache/hadoop/hive/ql/plan/FileSinkDesc.java (hunk 6 above)
Comment (pvargacl) [reply ql/src/java/org/apache/hadoop/hive/ql/plan/FileSinkDesc.java:114]: done

File: ql/src/java/org/apache/hadoop/hive/ql/exec/Utilities.java (hunk 5 above)
Comment (pvargacl) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/Utilities.java:4531]: At the end it is used as a List in PartitionDetails, I could use a set here, but I would convert it anyway to a list.

**Review by deniskuzZ — APPROVED** [review]
LGTM

File: ql/src/java/org/apache/hadoop/hive/ql/exec/Utilities.java (hunk 5 above)
Comment (pvary) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/Utilities.java:4531]: What is the expected order of the items in the list?
Do we really need a list there?

Comment (pvargacl) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/Utilities.java:4531]: Not really, but it starts a refactor domino. I started it but reverted, I do not want to pollute this PR.

**Review by pvary — APPROVED** [review]

**Pull request merged**